File: test/os_stat.c
/* Behavioral contract for the file metadata kernels behind `os/stat` and
 * `os/lstat`, run against whichever implementation the build selected
 * (`-Dos-stat=c` or the Zig default). */

#include <assert.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <janet.h>

#ifdef JANET_WINDOWS
#include <direct.h>
#else
#include <unistd.h>
#endif

const char *janet_os_mode_name(uint32_t mode);
int32_t janet_os_decode_permissions(uint32_t mode);
int32_t janet_os_perm_to_unix(uint32_t mode);
uint32_t janet_os_perm_from_unix(int32_t permissions);
int32_t janet_os_stat_field_count(void);
const char *janet_os_stat_field_name(int32_t index);
int32_t janet_os_stat_field_lookup(const uint8_t *key, int32_t len);

/* The registry's order is also the field identifier the C getters switch on,
 * so both are pinned here. */
static const char *const expected_fields[] = {
    "dev",
    "inode",
    "mode",
    "int-permissions",
    "permissions",
    "uid",
    "gid",
    "nlink",
    "rdev",
    "size",
    "blocks",
    "blocksize",
    "accessed",
    "modified",
    "changed"
};

static const char work_dir[] = "janet-zig-os-stat-4d71";
static const char work_file[] = "janet-zig-os-stat-4d71/file";
static const char work_link[] = "janet-zig-os-stat-4d71/link";

static void expect_mode_name(uint32_t mode, const char *expected) {
    assert(!strcmp(janet_os_mode_name(mode), expected));
}

static void test_mode_names(void) {
#ifdef JANET_WINDOWS
    /* The CRT has no S_IS* macros, and Janet tests the three type bits
     * individually rather than masking first. */
    expect_mode_name(_S_IFREG | 0666, "file");
    expect_mode_name(_S_IFDIR | 0777, "directory");
    expect_mode_name(_S_IFCHR, "character");
    expect_mode_name(0, "other");
    expect_mode_name(0644, "other");
#else
    expect_mode_name(S_IFREG | 0644, "file");
    expect_mode_name(S_IFDIR | 0755, "directory");
    expect_mode_name(S_IFCHR | 0666, "character");
    expect_mode_name(0, "other");
    expect_mode_name(0777, "other");
#ifndef JANET_PLAN9
    expect_mode_name(S_IFIFO | 0644, "fifo");
    expect_mode_name(S_IFBLK | 0644, "block");
    expect_mode_name(S_IFSOCK | 0644, "socket");
    expect_mode_name(S_IFLNK | 0777, "link");
#endif
#endif
}

static void test_permission_bits(void) {
#ifdef JANET_WINDOWS
    assert(janet_os_decode_permissions(_S_IFREG | 0777) == (S_IEXEC | S_IWRITE | S_IREAD));
    assert(janet_os_perm_to_unix(S_IREAD) == 0444);
    assert(janet_os_perm_to_unix(S_IWRITE) == 0222);
    assert(janet_os_perm_to_unix(S_IEXEC) == 0111);
    assert(janet_os_perm_to_unix(S_IREAD | S_IWRITE | S_IEXEC) == 0777);
    /* Windows collapses user, group, and other, so only the reduced value
     * round-trips. */
    assert(janet_os_perm_to_unix(janet_os_perm_from_unix(0777)) == 0777);
#else
    /* Type bits and the setuid/setgid/sticky field are dropped. */
    assert(janet_os_decode_permissions(S_IFREG | 0754) == 0754);
    assert(janet_os_decode_permissions(S_IFDIR | 07777) == 0777);
    assert(janet_os_decode_permissions(0) == 0);

    /* The portable value and the host's mode agree on Unix. */
    for (int32_t mode = 0; mode <= 0777; mode++) {
        assert(janet_os_perm_to_unix((uint32_t) mode) == mode);
        assert(janet_os_perm_from_unix(mode) == (uint32_t) mode);
        assert(janet_os_perm_to_unix(janet_os_perm_from_unix(mode)) == mode);
    }
#endif
}

static void test_field_registry(void) {
    int32_t count = janet_os_stat_field_count();
    assert(count == (int32_t)(sizeof(expected_fields) / sizeof(expected_fields[0])));

    for (int32_t index = 0; index < count; index++) {
        const char *name = janet_os_stat_field_name(index);
        assert(name != NULL);
        assert(!strcmp(name, expected_fields[index]));
        assert(janet_os_stat_field_lookup((const uint8_t *) name, (int32_t) strlen(name)) == index);
    }

    /* Out-of-range indices report absence rather than reading past the table. */
    assert(janet_os_stat_field_name(-1) == NULL);
    assert(janet_os_stat_field_name(count) == NULL);

    /* Lookup matches whole names only. */
    assert(janet_os_stat_field_lookup((const uint8_t *) "de", 2) == -1);
    assert(janet_os_stat_field_lookup((const uint8_t *) "device", 6) == -1);
    assert(janet_os_stat_field_lookup((const uint8_t *) "", 0) == -1);
    assert(janet_os_stat_field_lookup((const uint8_t *) "dev", -1) == -1);
    assert(janet_os_stat_field_lookup((const uint8_t *) "Dev", 3) == -1);
    assert(janet_os_stat_field_lookup((const uint8_t *) "int-permission", 14) == -1);

    /* `janet_cstrcmp`, which this lookup replaced, stops at a NUL the key and
     * the name share, so a key whose own bytes end in NUL still matches. That
     * quirk is preserved deliberately. */
    assert(janet_os_stat_field_lookup((const uint8_t *) "dev\0", 4) == 0);
}

static void run(JanetTable *env, const char *source) {
    Janet result;
    assert(janet_dostring(env, source, "os-stat-contract", &result) == 0);
}

static void test_core_functions(void) {
    JanetTable *env = janet_core_env(NULL);

    run(env,
        "(os/mkdir \"janet-zig-os-stat-4d71\")\n"
        "(spit \"janet-zig-os-stat-4d71/file\" \"0123456789\")\n"
        "(os/chmod \"janet-zig-os-stat-4d71/file\" 8r640)\n");

    /* A whole-table result carries every registry field, in the registry's
     * own names. */
    run(env,
        "(def st (os/stat \"janet-zig-os-stat-4d71/file\"))\n"
        "(assert (table? st))\n"
        "(assert (= 15 (length st)))\n"
        "(each key [:dev :inode :mode :int-permissions :permissions :uid :gid\n"
        "           :nlink :rdev :size :blocks :blocksize :accessed :modified :changed]\n"
        "  (assert (not= nil (st key))))\n");

    /* Classification, size, and both permission projections. */
    run(env,
        "(def st (os/stat \"janet-zig-os-stat-4d71/file\"))\n"
        "(assert (= :file (st :mode)))\n"
        "(assert (= 10 (st :size)))\n"
        "(assert (= :directory (get (os/stat \"janet-zig-os-stat-4d71\") :mode)))\n");
#ifndef JANET_WINDOWS
    run(env,
        "(def st (os/stat \"janet-zig-os-stat-4d71/file\"))\n"
        "(assert (= 8r640 (st :int-permissions)))\n"
        "(assert (= \"rw-r-----\" (st :permissions)))\n");
#endif

    /* A keyword selects one field; an unknown keyword raises. */
    run(env,
        "(assert (= :file (os/stat \"janet-zig-os-stat-4d71/file\" :mode)))\n"
        "(assert (= 10 (os/stat \"janet-zig-os-stat-4d71/file\" :size)))\n"
        "(assert (= :directory (os/stat \"janet-zig-os-stat-4d71\" :mode)))\n"
        "(assert (not (first (protect (os/stat \"janet-zig-os-stat-4d71/file\" :nope)))))\n"
        "(assert (not (first (protect (os/stat \"janet-zig-os-stat-4d71/file\" :de)))))\n");

    /* A supplied table is filled and returned. */
    run(env,
        "(def tab @{:seed true})\n"
        "(assert (= tab (os/stat \"janet-zig-os-stat-4d71/file\" tab)))\n"
        "(assert (= 16 (length tab)))\n"
        "(assert (= :file (tab :mode)))\n");

    /* A missing path is nil rather than an error. */
    run(env, "(assert (nil? (os/stat \"janet-zig-os-stat-4d71/missing\")))\n");

#if !defined(JANET_WINDOWS) && !defined(JANET_NO_SYMLINKS)
    /* `os/lstat` reports the link itself, `os/stat` its target. */
    run(env,
        "(os/symlink \"file\" \"janet-zig-os-stat-4d71/link\")\n"
        "(assert (= :link (os/lstat \"janet-zig-os-stat-4d71/link\" :mode)))\n"
        "(assert (= :file (os/stat \"janet-zig-os-stat-4d71/link\" :mode)))\n"
        "(assert (= 10 (os/stat \"janet-zig-os-stat-4d71/link\" :size)))\n");
#endif
}

static void clean_paths(void) {
    (void) remove(work_link);
    (void) remove(work_file);
#ifdef JANET_WINDOWS
    (void) _rmdir(work_dir);
#else
    (void) rmdir(work_dir);
#endif
}

void os_stat_contract(void) {
    test_mode_names();
    test_permission_bits();
    test_field_registry();

    clean_paths();
    janet_init();
    test_core_functions();
    janet_deinit();
    clean_paths();
}

File: test/os_stat.zig
//! Behavioral contract for the file-metadata kernels behind `os/stat` and
//! `os/lstat`: mode classification, the two permission projections, and the
//! field registry.
//!
//! ## The registry's order is an identifier, not a presentation choice
//!
//! `os_stat.fieldName(i)` and `os_stat.fieldLookup(name)` are
//! inverses, and the index they agree on is what the getters switch on. So the
//! *order* of the fifteen names is load-bearing: inserting a field in the
//! middle renumbers everything after it, and every getter then reads the wrong
//! member. Nothing in Janet can see an index, `os/stat` taking keywords, so
//! the list below is the only place that ordering is written down twice.
//!
//! ## Two permission projections, and only one of them round-trips everywhere
//!
//! `os_stat.hostPermToUnix` and `os_stat.hostPermFromUnix` convert between the
//! host's mode bits and Janet's portable nine. On Unix they are the identity
//! and every one of the 512 values round-trips, which is asserted
//! exhaustively. On Windows the CRT collapses user, group and other into three
//! bits, so only the reduced value survives a round trip, and the contract is
//! weaker there because the platform is, and saying which is the point.
//!
//! ## One preserved quirk
//!
//! `os_stat.fieldLookup` keeps `utils.cstrcmp`'s rule, which stops at a NUL
//! shared by the key and the name. So a key whose own bytes end in NUL
//! still matches: `lookup("dev\0", 4)` is 0 rather than -1. That is deliberate
//! and is asserted below so a port cannot tidy it away.

// ==========================================================================
// Standard library imports
// ==========================================================================

const std = @import("std");
const builtin = @import("builtin");

// ==========================================================================
// Project imports
// ==========================================================================

const config = @import("config");
const core_env = @import("subsystems").env;
const expect = @import("expect.zig").expect;
const fs = @import("subsystems").fs;
const harness = @import("harness.zig");

/// The field registry, by import.
const os_stat = @import("subsystems").stat;
const repr = @import("repr");
const tables = @import("subsystems").value.tables;
const value = @import("subsystems").value;
const vm_lifecycle = @import("subsystems").lifecycle;

// ==========================================================================
// Constants
// ==========================================================================

var environment: *tables.Table = undefined;

/// The registry, in order. See the header comment on why the order matters.
const expected_fields = [_][]const u8{
    "dev",
    "inode",
    "mode",
    "int-permissions",
    "permissions",
    "uid",
    "gid",
    "nlink",
    "rdev",
    "size",
    "blocks",
    "blocksize",
    "accessed",
    "modified",
    "changed",
};

const work_dir = "janet-zig-os-stat-4d71";
const work_file = "janet-zig-os-stat-4d71/file";
const work_link = "janet-zig-os-stat-4d71/link";

// ==========================================================================
// Cases
// ==========================================================================

fn modeNameIs(mode: u32, expected: []const u8) bool {
    return std.mem.eql(u8, std.mem.span(os_stat.hostModeName(mode)), expected);
}

fn eval(source: [*:0]const u8) void {
    var result: repr.Value = undefined;
    expect(core_env.dostring(environment, source, "os-stat-contract", &result) == 0);
}

fn cleanPaths() void {
    _ = fs.hostRemove(work_link);
    _ = fs.hostRemove(work_file);
    _ = fs.hostRmdir(work_dir);
}

/// Janet tests the three type bits individually rather than masking with
/// `S_IFMT` first, so a mode with no type bits at all is "other"
/// rather than a misclassification.
///
/// Unix only, and the guard is a `comptime` condition so the branch is not
/// analysed on a Windows cross-compile, `std.c.S` having no members there.
/// There is no CRT arm beside it, because this tree's platform
/// scope makes Windows a build target rather than a tested one, so the arm is
/// dropped rather than written and never run.
fn theModeNames() void {
    if (builtin.os.tag == .windows) return;
    const S = std.c.S;

    expect(modeNameIs(S.IFREG | 0o644, "file"));
    expect(modeNameIs(S.IFDIR | 0o755, "directory"));
    expect(modeNameIs(S.IFCHR | 0o666, "character"));
    expect(modeNameIs(0, "other"));
    expect(modeNameIs(0o777, "other"));

    // Plan 9 has none of these four.
    if (!(builtin.os.tag == .plan9)) {
        expect(modeNameIs(S.IFIFO | 0o644, "fifo"));
        expect(modeNameIs(S.IFBLK | 0o644, "block"));
        expect(modeNameIs(S.IFSOCK | 0o644, "socket"));
        expect(modeNameIs(S.IFLNK | 0o777, "link"));
    }
}

/// Unix only, for the same reason as `theModeNames`. Windows collapses user,
/// group and other into three CRT bits, so only the reduced value round-trips
/// there and the exhaustive sweep below would be wrong rather than merely
/// unavailable.
fn thePermissionBits() void {
    if (builtin.os.tag == .windows) return;
    const S = std.c.S;

    // Type bits and the setuid/setgid/sticky field are dropped.
    expect(os_stat.hostDecodePermissions(S.IFREG | 0o754) == 0o754);
    expect(os_stat.hostDecodePermissions(S.IFDIR | 0o7777) == 0o777);
    expect(os_stat.hostDecodePermissions(0) == 0);

    // On Unix the two projections are the identity, and all 512 round-trip.
    var mode: i32 = 0;
    while (mode <= 0o777) : (mode += 1) {
        expect(os_stat.hostPermToUnix(@intCast(mode)) == mode);
        expect(os_stat.hostPermFromUnix(mode) == @as(u32, @intCast(mode)));
        expect(os_stat.hostPermToUnix(os_stat.hostPermFromUnix(mode)) == mode);
    }
}

fn theFieldRegistry() void {
    const count = os_stat.fieldCount();
    expect(count == expected_fields.len);

    for (expected_fields, 0..) |expected, index| {
        const name = os_stat.fieldName(@intCast(index)).?;
        const actual = std.mem.span(name);
        expect(std.mem.eql(u8, actual, expected));
        // The inverse agrees, which is what makes the index an identifier.
        expect(os_stat.fieldLookup(actual.ptr, @intCast(actual.len)) == index);
    }

    // Out of range reports absence rather than reading past the table.
    expect(os_stat.fieldName(-1) == null);
    expect(os_stat.fieldName(count) == null);

    // Whole names only: no prefix, no extension, no empty key, and no case
    // folding. The negative length C also had to refuse is not a value this
    // signature admits: `len` is a byte count and is a `usize`.
    expect(os_stat.fieldLookup("de", 2) == -1);
    expect(os_stat.fieldLookup("device", 6) == -1);
    expect(os_stat.fieldLookup("", 0) == -1);
    expect(os_stat.fieldLookup("Dev", 3) == -1);
    expect(os_stat.fieldLookup("int-permission", 14) == -1);

    // The preserved `utils.cstrcmp` quirk; see the header comment.
    expect(os_stat.fieldLookup("dev\x00", 4) == 0);
}

/// Written in Janet because every assertion here is about a *keyword-keyed
/// table*, which Janet states in a line and Zig states in five unwraps.
fn theCoreFunctions() void {
    eval(
        \\(os/mkdir "janet-zig-os-stat-4d71")
        \\(spit "janet-zig-os-stat-4d71/file" "0123456789")
        \\(os/chmod "janet-zig-os-stat-4d71/file" 8r640)
    );

    // A whole-table result has every registry field in it, under the
    // registry's own names, which is the Janet-visible half of
    // `theFieldRegistry`.
    eval(
        \\(def st (os/stat "janet-zig-os-stat-4d71/file"))
        \\(assert (table? st))
        \\(assert (= 15 (length st)))
        \\(each key [:dev :inode :mode :int-permissions :permissions :uid :gid
        \\           :nlink :rdev :size :blocks :blocksize :accessed :modified :changed]
        \\  (assert (not= nil (st key))))
    );

    eval(
        \\(def st (os/stat "janet-zig-os-stat-4d71/file"))
        \\(assert (= :file (st :mode)))
        \\(assert (= 10 (st :size)))
        \\(assert (= :directory (get (os/stat "janet-zig-os-stat-4d71") :mode)))
    );

    // Unix only: the CRT has three permission bits rather than nine.
    if (builtin.os.tag != .windows) eval(
        \\(def st (os/stat "janet-zig-os-stat-4d71/file"))
        \\(assert (= 8r640 (st :int-permissions)))
        \\(assert (= "rw-r-----" (st :permissions)))
    );

    // A keyword selects one field rather than building the table.
    eval(
        \\(assert (= :file (os/stat "janet-zig-os-stat-4d71/file" :mode)))
        \\(assert (= 10 (os/stat "janet-zig-os-stat-4d71/file" :size)))
        \\(assert (= :directory (os/stat "janet-zig-os-stat-4d71" :mode)))
    );

    // A supplied table is filled and returned: the same table, with its
    // existing entry intact, so the length is sixteen.
    eval(
        \\(def tab @{:seed true})
        \\(assert (= tab (os/stat "janet-zig-os-stat-4d71/file" tab)))
        \\(assert (= 16 (length tab)))
        \\(assert (= :file (tab :mode)))
    );

    // A missing path is nil rather than an error, which distinguishes "no such
    // file" from "the call failed".
    eval("(assert (nil? (os/stat \"janet-zig-os-stat-4d71/missing\")))");

    if (config.symlinks) {
        // `os/lstat` reports the link, `os/stat` its target.
        eval(
            \\(os/symlink "file" "janet-zig-os-stat-4d71/link")
            \\(assert (= :link (os/lstat "janet-zig-os-stat-4d71/link" :mode)))
            \\(assert (= :file (os/stat "janet-zig-os-stat-4d71/link" :mode)))
            \\(assert (= 10 (os/stat "janet-zig-os-stat-4d71/link" :size)))
        );
    }
}

/// An unknown field keyword raises rather than giving nil, and a prefix of a
/// real one is still unknown. These are reached
/// inside a Janet string; here they are values.
fn theRefusals() void {
    const stat = harness.core("os/stat");
    var args: [2]repr.Value = undefined;

    args[0] = value.fromBytes(work_file, .string);
    args[1] = value.fromBytes("nope", .keyword);
    expect(harness.raised(stat, .{args[0..2]}) != null);

    // `:de` is a prefix of `:dev`, and the lookup matches whole names.
    args[1] = value.fromBytes("de", .keyword);
    expect(harness.raised(stat, .{args[0..2]}) != null);
}

// ==========================================================================
// Entry
// ==========================================================================

pub fn run() void {
    theModeNames();
    thePermissionBits();
    theFieldRegistry();

    cleanPaths();
    harness.init();
    environment = harness.coreEnv();
    theCoreFunctions();
    theRefusals();
    vm_lifecycle.deinit();
    cleanPaths();
}
